src/lib.rs

```diff
@@ -227,41 +227,78 @@ pub mod pallet {
-		pub fn my_first_claim(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+		/// Collect whatever portion of your reward are currently vested. The first time each
+		/// contributor calls this function pays no fees
+		#[pallet::weight(0)]
+		pub fn show_me_the_money(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
 			let payee = ensure_signed(origin)?;
-
-			// Calculate the vested amount on demand.
-			let info = AccountsPayable::<T>::get(&payee).ok_or(Error::<T>::NoAssociatedClaim)?;
-
+			let initialized = <Initialized<T>>::get();
 			ensure!(
-				info.free_claim_done == false,
-				Error::<T>::FirstClaimAlreadyDone
+				initialized == true,
+				Error::<T>::RewardVecNotFullyInitializedYet
 			);
+			// Calculate the veted amount on demand.
+			let mut info =
+				AccountsPayable::<T>::get(&payee).ok_or(Error::<T>::NoAssociatedClaim)?;
 			ensure!(
 				info.claimed_reward < info.total_reward,
 				Error::<T>::RewardsAlreadyClaimed
 			);
 
-			Self::make_vested_payment(info, payee.clone(), true)
-		}
+			// Vesting is done in relation with the relay chain slot
+			let now: T::BlockNumber =
+				cumulus_pallet_parachain_system::Pallet::<T>::validation_data()
+					.expect("validation data was set in parachain system inherent")
+					.relay_parent_number
+					.into();
 
-		/// Collect whatever portion of your reward are currently vested.
-		#[pallet::weight(0)]
-		pub fn show_me_the_money(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
-			let payee = ensure_signed(origin)?;
+			// Substract the first payment from the vested amount
+			let first_paid = T::InitializationPayment::get() * info.total_reward;
 
-			// Calculate the veted amount on demand.
-			let info = AccountsPayable::<T>::get(&payee).ok_or(Error::<T>::NoAssociatedClaim)?;
-			ensure!(
-				info.claimed_reward < info.total_reward,
-				Error::<T>::RewardsAlreadyClaimed
-			);
+			// Vesting perdiods as balance
+			let vesting_period = T::VestingPeriod::get()
+				.saturated_into::<u128>()
+				.try_into()
+				.ok()
+				.ok_or(Error::<T>::WrongConversionU128ToBalance)?;
 
-			Self::make_vested_payment(info, payee.clone(), false)
-		}
+			// To calculate how much could the user have claimed already
+			let payable_period = now.saturating_sub(<InitRelayBlock<T>>::get().into());
 
+			let pay_period_as_balance: BalanceOf<T> = payable_period
+				.saturated_into::<u128>()
+				.try_into()
+				.ok()
+				.ok_or(Error::<T>::WrongConversionU128ToBalance)?;
+
+			// How much should the contributor have already claimed by this block?
+			// By multiplying first we allow the conversion to integer done with the biggest number
+			let should_have_claimed = pay_period_as_balance
+				.saturating_mul(info.total_reward - first_paid)
+				/ vesting_period;
+
+			// If the period is bigger than whats missing to pay, then return whats missing to pay
+			let payable_amount = if should_have_claimed >= info.total_reward {
+				info.total_reward.saturating_sub(info.claimed_reward)
```

Comment:
This case is supposed to be when the entire amount is vested but we haven't claimed it all, right? Is this going to be slightly off because we subtract `first_paid` above?